src/Our.Umbraco.Vorto/Web/Controllers/VortoApiController.cs

```diff
@@ -99,36 +99,42 @@ public IEnumerable<object> GetLanguages(string section, int id, int parentId, Gu
 
 			if (languageSource == "inuse")
 			{
-				var xpath = preValues.ContainsKey("xpath") ? preValues["xpath"].Value : "";
-
-				// Grab languages by xpath (only if in content section)
-                if (!string.IsNullOrWhiteSpace(xpath) && section == "content")
-				{
-					xpath = xpath.Replace("$currentPage",
-						string.Format("//*[@id={0} and @isDoc]", id)).Replace("$parentPage",
-							string.Format("//*[@id={0} and @isDoc]", parentId)).Replace("$ancestorOrSelf",
-								string.Format("//*[@id={0} and @isDoc]", id != 0 ? id : parentId));
-
-					// Lookup language nodes
-					var nodeIds = uQuery.GetNodesByXPath(xpath).Select(x => x.Id).ToArray();
-					if (nodeIds.Any())
-					{
-						var db = ApplicationContext.Current.DatabaseContext.Database;
-						languages.AddRange(db.Query<string>(
-							string.Format(
-								"SELECT DISTINCT [languageISOCode] FROM [umbracoLanguage] JOIN [umbracoDomains] ON [umbracoDomains].[domainDefaultLanguage] = [umbracoLanguage].[id] WHERE [umbracoDomains].[domainRootStructureID] in ({0})",
-								string.Join(",", nodeIds)))
-							.Select(CultureInfo.GetCultureInfo)
-							.Select(x => new Language
-							{
-								IsoCode = x.Name,
-								Name = x.DisplayName,
-								NativeName = x.NativeName,
+                var currentNode = id != 0 ? ApplicationContext.Services.ContentService.GetById(id) : null;
```

Comment:
I'd be tempted to refactor out the use of the ContentService here as this will be performing a database hit. You could probably go one of two approaches. 1) given you are really only checking to see if the content is actually published, you could check the published content cache and if it's not their, assume it's unpublished or 2) perform a lucene query on the internal lucene index which contains unpublished nodes and check there. I reckon option 1 would be easier but should do the trick I think.

Reply:
Great, thanks for taking a look Matt - I'll update and re-commit just now :)